src/eos/chemTab.cpp

```diff
@@ -452,23 +528,45 @@ void ablate::eos::ChemTab::ChemTabSourceCalculator::AddSource(const ablate::doma
     VecGetArrayRead(locX, &xArray) >> utilities::PetscUtilities::checkError;
 
     // Get the solution dm
-    DM dm;
+    DM dm;  // NOTE: DM is topological space (i.e. grid)
     VecGetDM(locFVec, &dm) >> utilities::PetscUtilities::checkError;
 
+    // Here we store the batched pointers needed for multiple chemTabModel->ChemistrySource() calls
+    PetscInt buffer_len = cellRange.end - cellRange.start;
+    PetscScalar allDensity[buffer_len];
+    const PetscScalar *allDensityCPV[buffer_len];
+    PetscScalar *allDensityEnergySource[buffer_len];
+    PetscScalar *allDensityCPVSource[buffer_len];
+    // PetscScalar* allSourceAtCell[buffer_len]; // apparently these can't be used b/c of secret offsets
+    // const PetscScalar* allSolutionAtCell[buffer_len]; // apparently these can't be used b/c of secret offsets
+
+    // NOTE: this clunky approach has been verified to work for batch processsing
     // March over each cell in the range
     for (PetscInt c = cellRange.start; c < cellRange.end; ++c) {
         const PetscInt iCell = cellRange.points ? cellRange.points[c] : c;
+        // Dwyer: iCell is the "point"
 
-        // Get the current state variables for this cell
         PetscScalar *sourceAtCell = nullptr;
         DMPlexPointLocalRef(dm, iCell, fArray, &sourceAtCell) >> utilities::PetscUtilities::checkError;
 
-        // Get the current state variables for this cell
+        // Get the current state variables for this cell (CPVs)
         const PetscScalar *solutionAtCell = nullptr;
         DMPlexPointLocalRead(dm, iCell, xArray, &solutionAtCell) >> utilities::PetscUtilities::checkError;
-
-        chemTabModel->ChemistrySource(solutionAtCell[densityOffset], solutionAtCell + densityProgressVariableOffset, sourceAtCell + densityEnergyOffset, sourceAtCell + densityProgressVariableOffset);
+        // NOTE: DMPlexPointLocalRef() is Read/Write while DMPlexPointLocalRead() is Read only
+
+        // store this cell's attributes into arg vectors
+        size_t index = c - cellRange.start;
+        allDensity[index] = solutionAtCell[densityOffset];
+        allDensityCPV[index] = solutionAtCell + densityProgressVariableOffset;
+        allDensityEnergySource[index] = sourceAtCell + densityEnergyOffset;
+        allDensityCPVSource[index] = sourceAtCell + densityProgressVariableOffset;
     }
+
+    // Using batch overloaded version
+    // TODO: consider/optimize the problem that is likely requires loop to copy these arrays into TF inputs??
```

Comment:
If the ChemTabSourceCalculator instance owned the tensor, you could copy directly into the tensor here.